go/src/lib.rs

```diff
@@ -1,9 +1,172 @@
+pub type SuccessCallback =
+    unsafe extern "C" fn(channel_address: usize, message: *const c_char) -> ();
+
+/// Failure callback that is called when a Redis command fails.
+///
+/// `error` should be manually freed by calling `free_error` after this callback is invoked, otherwise a memory leak will occur.
+pub type FailureCallback =
+    unsafe extern "C" fn(channel_address: usize, error: *const RedisErrorFFI) -> ();
+
+/// The connection response.
+///
+/// It contains either a connection or an error. It is represented as a struct instead of an enum for ease of use in the wrapper language.
+#[repr(C)]
+pub struct ConnectionResponse {
+    conn_ptr: *const c_void,
+    error: *const RedisErrorFFI,
+}
+
+/// A Redis error.
+#[repr(C)]
+pub struct RedisErrorFFI {
+    message: *const c_char,
+    error_type: ErrorType,
+}
+
+/// FFI compatible version of the ErrorType enum defined in protobuf.
+// TODO: Update when command errors are implemented
+#[repr(C)]
+pub enum ErrorType {
+    ClosingError = 0,
+    RequestError = 1,
+    TimeoutError = 2,
+    ExecAbortError = 3,
+    ConnectionError = 4,
+}
+
+/// The glide client.
+#[allow(dead_code)]
+pub struct Client {
+    client: GlideClient,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+    runtime: Runtime,
+}
+
+fn create_client_internal(
+    connection_request_bytes: &[u8],
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> Result<Client, RedisErrorFFI> {
+    let request = connection_request::ConnectionRequest::parse_from_bytes(connection_request_bytes)
+        .map_err(|err| {
+            let msg = CString::new(err.to_string()).unwrap();
+            RedisErrorFFI {
+                message: msg.into_raw(),
+                error_type: ErrorType::ConnectionError,
+            }
+        })?;
+    let runtime = Builder::new_multi_thread()
+        .enable_all()
+        .thread_name("GLIDE for Redis Go thread")
+        .build()
+        .map_err(|err| {
+            let msg = CString::new(err.to_string()).unwrap();
+            RedisErrorFFI {
+                message: msg.into_raw(),
+                error_type: ErrorType::ConnectionError,
+            }
+        })?;
+    let _runtime_handle = runtime.enter();
+    let client = runtime.block_on(GlideClient::new(request)).map_err(|err| {
+        let msg = CString::new(err.to_string()).unwrap();
+        RedisErrorFFI {
+            message: msg.into_raw(),
+            error_type: ErrorType::ConnectionError,
+        }
+    })?;
+    Ok(Client {
+        client,
+        success_callback,
+        failure_callback,
+        runtime,
+    })
+}
+
+/// Creates a new client to the given address. The success callback needs to copy the given string synchronously, since it will be dropped by Rust once the callback returns. All callbacks should be offloaded to separate threads in order not to exhaust the client's thread pool.
```

Comment:
```suggestion
/// Creates a new client with the given configuration. The success callback needs to copy the given string synchronously, since it will be dropped by Rust once the callback returns. All callbacks should be offloaded to separate threads in order not to exhaust the client's thread pool.
```